### PR TITLE
algorithm: implement polynomial regression

### PR DESCRIPTION
**Pull request by amo33**
This PR adds regression module and it's unit test.
Output using `make test` :
```
    Start 2: RegressionTest.LinearRegression
2/3 Test #2: RegressionTest.LinearRegression .......   Passed    0.42 sec
    Start 3: RegressionTest.PolynomialRegression
3/3 Test #3: RegressionTest.PolynomialRegression ...   Passed    0.55 sec

100% tests passed, 0 tests failed out of 3

Total Test time (real) =   1.00 sec
```

Test Log :
```
2/3 Testing: RegressionTest.LinearRegression
2/3 Test: RegressionTest.LinearRegression
Command: "/home/flatflow/build/flatflow/scheduler/internal/algorithm/algorithm_test" "--gtest_filter=RegressionTest.LinearRegression"
Directory: /home/flatflow/build/flatflow/scheduler/internal/algorithm
"RegressionTest.LinearRegression" start time: Apr 23 07:16 UTC
Output:
----------------------------------------------------------
Running main() from /home/flatflow/third_party/googletest/googletest/src/gtest_main.cc
Note: Google Test filter = RegressionTest.LinearRegression
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from RegressionTest
[ RUN      ] RegressionTest.LinearRegression
[       OK ] RegressionTest.LinearRegression (413 ms)
[----------] 1 test from RegressionTest (413 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (413 ms total)
[  PASSED  ] 1 test.
<end of output>
Test time =   0.42 sec
----------------------------------------------------------
Test Passed.
"RegressionTest.LinearRegression" end time: Apr 23 07:16 UTC
"RegressionTest.LinearRegression" time elapsed: 00:00:00
----------------------------------------------------------

3/3 Testing: RegressionTest.PolynomialRegression
3/3 Test: RegressionTest.PolynomialRegression
Command: "/home/flatflow/build/flatflow/scheduler/internal/algorithm/algorithm_test" "--gtest_filter=RegressionTest.PolynomialRegression"
Directory: /home/flatflow/build/flatflow/scheduler/internal/algorithm
"RegressionTest.PolynomialRegression" start time: Apr 23 07:16 UTC
Output:
----------------------------------------------------------
Running main() from /home/flatflow/third_party/googletest/googletest/src/gtest_main.cc
Note: Google Test filter = RegressionTest.PolynomialRegression
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from RegressionTest
[ RUN      ] RegressionTest.PolynomialRegression
[       OK ] RegressionTest.PolynomialRegression (544 ms)
[----------] 1 test from RegressionTest (544 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (544 ms total)
[  PASSED  ] 1 test.
<end of output>
Test time =   0.55 sec
----------------------------------------------------------
Test Passed.
"RegressionTest.PolynomialRegression" end time: Apr 23 07:16 UTC
"RegressionTest.PolynomialRegression" time elapsed: 00:00:00
----------------------------------------------------------

End testing: Apr 23 07:16 UTC

```

### CODE REVIEW

File: flatflow/scheduler/internal/algorithm/regression.h [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+// See _sgd_fast.pyx.tp for detail implementation.
+//
+// References:
+// https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.PassiveAggressiveRegressor.html#sklearn.linear_model.PassiveAggressiveRegressor
+// https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/linear_model/_sgd_fast.pyx.tp
+template <std::size_t degree>
+class PassiveAggressiveRegressor {
+ public:
+  inline explicit PassiveAggressiveRegressor(const double &epsilon)
+      : epsilon_(epsilon), coef_(degree + 1, 1.0), power_(degree + 1, 0.0) {
```
Comment (9rum) [inline flatflow/scheduler/internal/algorithm/regression.h:52]: Please put only simple assignments in initializer list, not constructor calling.
That is, initialization of ``coef_`` and ``power_`` should be placed in the constructor body.

File: flatflow/scheduler/internal/algorithm/regression_test.cc [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+#include "absl/base/log_severity.h"
+#include "absl/log/globals.h"
+#include "absl/log/initialize.h"
+#include "absl/log/internal/globals.h"
+#include "absl/log/log.h"
+#include "gtest/gtest.h"
+
+namespace {
+
+template <int degree>
```
Comment (9rum) [inline flatflow/scheduler/internal/algorithm/regression_test.cc:29]: The template parameter ``degree`` should be ``std::size_t``.

File: flatflow/scheduler/internal/algorithm/regression.h [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FLATFLOW_SCHEDULER_INTERNAL_ALGORITHM_REGRESSION_H_
+#define FLATFLOW_SCHEDULER_INTERNAL_ALGORITHM_REGRESSION_H_
+
+#include <omp.h>
```
Comment (9rum) [inline flatflow/scheduler/internal/algorithm/regression.h:18]: It seems there is no use of OpenMP in this file, so this ``include`` statement should be omitted.

File: flatflow/scheduler/internal/algorithm/regression.h [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  // This implementation is based on Epsilon-Insensitive Passive Aggressive Regressor.
+  // This method is implementation of partialfit method in order to use in
+  // online learning.
+  inline void fit(const std::vector<double> &X, const std::vector<double> &y) {
+    CHECK(X.size() == y.size());
+    for (std::size_t iter = 0; iter < kMaxIter_; ++iter) {
+      for (std::size_t dataIdx = 0; dataIdx < X.size(); ++dataIdx) {
+        const auto prediction = predict(X[dataIdx]);
+        const auto loss =
+            std::max(0.0, std::abs(y[dataIdx] - prediction) - epsilon_);
```
Comment (9rum) [inline flatflow/scheduler/internal/algorithm/regression.h:68]: Does it work doing ``std::max()`` with zero after ``std::abs()``?

File: flatflow/scheduler/internal/algorithm/regression.h [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        }
+      }
+    }
+  }
+
+  // PassiveAggressiveRegressor::predict()
+  //
+  // Predicts value based on given scalar x and current model coef_.
+  inline double predict(const double &x) const {
+    double prediction = 0.0;
```
Comment (9rum) [inline flatflow/scheduler/internal/algorithm/regression.h:89]: Use ``auto``.

File: flatflow/scheduler/internal/algorithm/regression_test.cc [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    std::random_device rd;
+    std::mt19937 gen(rd());
+    std::uniform_real_distribution<double> dis(300.0, 400.0);
+    for (std::size_t coefIdx = 0; coefIdx <= degree; ++coefIdx) {
+      coeffs.push_back(dis(gen));
+    }
+    return coeffs;
+  }
+
+  void print(const std::vector<double> &modelCofficients,
```
Comment (9rum) [inline flatflow/scheduler/internal/algorithm/regression_test.cc:78]: I think ``print`` is not feasible for this function name. ``print()`` does not print anything!

**Review by 9rum — CHANGES_REQUESTED** [review]
@amo33 Thanks for filing this PR.
I have left several comments, so please remind me after changes.
Also, please update ``flatflow/scheduler/internal/algorithm/CMakeLists.txt``.

File: flatflow/scheduler/internal/algorithm/regression.h (hunk 4 above)
Comment (amo33) [reply flatflow/scheduler/internal/algorithm/regression.h:68]: Yes. Loss is trying to get max value between an |y- prediction| and epsilon.
If epsilon is bigger than the |y-prediction| , loss becomes 0.0

File: flatflow/scheduler/internal/algorithm/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+if(BUILD_TESTING)
+  add_executable(
+    algorithm_test
+    regression_test.cc
```
Comment (9rum) [inline flatflow/scheduler/internal/algorithm/CMakeLists.txt:18]: There seems to have been some problems during upstream merge.
This file should contain exactly one line of changes.

File: flatflow/scheduler/internal/algorithm/regression.h [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  // PassiveAggressiveRegressor::fit()
+  //
+  // Fits the model to the given vector data X and y.
+  // This implementation is based on Epsilon-Insensitive Passive Aggressive
+  // Regressor. This method is implementation of partialfit method in order to
+  // use in online learning.
+  inline void fit(const std::vector<double> &X, const std::vector<double> &y) {
+    CHECK(X.size() == y.size());
```
Comment (9rum) [inline flatflow/scheduler/internal/algorithm/regression.h:63]: Use ``CHECK_EQ`` instead of ``CHECK``.

File: flatflow/scheduler/internal/algorithm/regression.h [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    }
+    return prediction;
+  }
+
+ protected:
+  double epsilon_;
+  std::vector<double> coef_;
+  std::vector<double> power_;
+  static constexpr auto kMaxIter_ = 1000;
+  static constexpr auto KC_ = 1.0;
```
Comment (9rum) [inline flatflow/scheduler/internal/algorithm/regression.h:101]: Fix typo: ``kC_`` instead of ``KC_``

**Review by 9rum — CHANGES_REQUESTED** [review]

**Comment by amo33** [comment]
@9rum Changes are updated.
Appreciate with detailed instructions.

**Review by 9rum — APPROVED** [review]
@amo33 Thanks for your hard work.
